File: trunk/Noxa.Emulation.Psp.Bios.FastHLE/Modules/Kernel_Library.cpp
// ----------------------------------------------------------------------------
// PSP Player Emulation Suite
// Copyright (C) 2006 Ben Vanik (noxa)
// Licensed under the LGPL - see License.txt in the project root for details
// ----------------------------------------------------------------------------

#include "Stdafx.h"
#define WIN32_LEAN_AND_MEAN
#include <Windows.h>

#include "Kernel_Library.h"
#include "Kernel.h"

using namespace System;
using namespace System::Diagnostics;
using namespace Noxa::Emulation::Psp;
using namespace Noxa::Emulation::Psp::Bios;
using namespace Noxa::Emulation::Psp::Bios::Modules;
using namespace Noxa::Emulation::Psp::Cpu;

void* Kernel_Library::QueryNativePointer( uint nid )
{
	//switch( nid )
	//{
	//case 0x092968F4:
	//case 0x5F10D406:
	//case 0x3B84732D:
	//case 0x47A0B729:
	//case 0xB55249D2:
	//}
	return 0;
}

// unsigned int sceKernelCpuSuspendIntr(); (/user/pspintrman.h:77)
int Kernel_Library::sceKernelCpuSuspendIntr()
{
	return _kernel->_core0->InterruptsFlag;
}

// void sceKernelCpuResumeIntr(unsigned int flags); (/user/pspintrman.h:84)
void Kernel_Library::sceKernelCpuResumeIntr( int flags )
{
	_kernel->_core0->InterruptsFlag = flags;
}

// void sceKernelCpuResumeIntrWithSync(unsigned int flags); (/user/pspintrman.h:91)
void Kernel_Library::sceKernelCpuResumeIntrWithSync( int flags )
{
	_kernel->_core0->InterruptsFlag = flags;
}

// int sceKernelIsCpuIntrSuspended(unsigned int flags); (/user/pspintrman.h:100)
int Kernel_Library::sceKernelIsCpuIntrSuspended( int flags )
{
	if( flags == 0x0 )
		return 1;
	else
		return 0;
}

// int sceKernelIsCpuIntrEnable(); (/user/pspintrman.h:107)
int Kernel_Library::sceKernelIsCpuIntrEnable()
{
	return _kernel->_core0->InterruptsEnabled;
}

File: trunk/Noxa.Emulation.Psp.Bios.ManagedHLE/Modules/Kernel_Library.cs
// ----------------------------------------------------------------------------
// PSP Player Emulation Suite
// Copyright (C) 2006 Ben Vanik (noxa)
// Licensed under the LGPL - see License.txt in the project root for details
// ----------------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Text;

using Noxa.Utilities;
using Noxa.Emulation.Psp;
using Noxa.Emulation.Psp.Bios;
using Noxa.Emulation.Psp.Cpu;

namespace Noxa.Emulation.Psp.Bios.ManagedHLE.Modules
{
	class Kernel_Library : Module
	{
		#region Properties

		public override string Name
		{
			get
			{
				return "Kernel_Library";
			}
		}

		#endregion

		#region State Management

		public Kernel_Library( Kernel kernel )
			: base( kernel )
		{
		}

		public override void Start()
		{
		}

		public override void Stop()
		{
		}

		#endregion

		// We should hope - for performance sake, that these
		// are implemented by the CPU natively

		[Stateless]
		[BiosFunction( 0x092968F4, "sceKernelCpuSuspendIntr" )]
		// SDK location: /user/pspintrman.h:77
		// SDK declaration: unsigned int sceKernelCpuSuspendIntr();
		public int sceKernelCpuSuspendIntr()
		{
			uint cpuFlags = _kernel.Cpu.InterruptsMask;
			_kernel.Cpu.InterruptsMask = 0;
			return ( int )cpuFlags;
		}

		[BiosFunction( 0x5F10D406, "sceKernelCpuResumeIntr" )]
		// SDK location: /user/pspintrman.h:84
		// SDK declaration: void sceKernelCpuResumeIntr(unsigned int flags);
		public void sceKernelCpuResumeIntr( int flags )
		{
			_kernel.Cpu.InterruptsMask = ( uint )flags;
		}

		[BiosFunction( 0x3B84732D, "sceKernelCpuResumeIntrWithSync" )]
		// SDK location: /user/pspintrman.h:91
		// SDK declaration: void sceKernelCpuResumeIntrWithSync(unsigned int flags);
		public void sceKernelCpuResumeIntrWithSync( int flags )
		{
			_kernel.Cpu.InterruptsMask = ( uint )flags;
		}

		[Stateless]
		[BiosFunction( 0x47A0B729, "sceKernelIsCpuIntrSuspended" )]
		// SDK location: /user/pspintrman.h:100
		// SDK declaration: int sceKernelIsCpuIntrSuspended(unsigned int flags);
		public int sceKernelIsCpuIntrSuspended( int flags )
		{
			uint cpuFlags = _kernel.Cpu.InterruptsMask;
			return ( ( cpuFlags & ( uint )flags ) != 0 ) ? 1 : 0;
		}

		[Stateless]
		[BiosFunction( 0xB55249D2, "sceKernelIsCpuIntrEnable" )]
		// SDK location: /user/pspintrman.h:107
		// SDK declaration: int sceKernelIsCpuIntrEnable();
		public int sceKernelIsCpuIntrEnable()
		{
			return ( _kernel.Cpu.InterruptsMask != 0 ) ? 1 : 0;
		}

	}
}

/* GenerateStubsV2: auto-generated - 57B381EB */
